nodeadm/internal/containerd/oci_config.go

```diff
@@ -0,0 +1,148 @@
+package containerd
+
+import (
+	"errors"
+	"io/fs"
+	"os/exec"
+	"reflect"
+	"slices"
+	"strings"
+
+	"github.com/awslabs/amazon-eks-ami/nodeadm/internal/api"
+	"github.com/awslabs/amazon-eks-ami/nodeadm/internal/util"
+	"github.com/pelletier/go-toml/v2"
+	"go.uber.org/zap"
+)
+
+type instanceTypeMixin struct {
+	instanceFamilies []string
+	apply            func(*[]byte) error
+}
+
+func (m *instanceTypeMixin) matches(cfg *api.NodeConfig) bool {
+	instanceFamily := strings.Split(cfg.Status.Instance.Type, ".")[0]
+	return slices.Contains(m.instanceFamilies, instanceFamily)
+}
+
+var (
+	// TODO: fetch this list dynamically
+	nvidiaInstances         = []string{"p3", "p3dn", "p4d", "p4de", "p5", "g4", "g4dn", "g5", "g6"}
+	neuronInstances         = []string{"inf1", "inf2", "trn1", "trn1n"}
+	NvidiaInstanceTypeMixin = instanceTypeMixin{
+		instanceFamilies: nvidiaInstances,
+		apply:            applyNvidia,
+	}
+
+	NeuronInstanceTypeMixin = instanceTypeMixin{
+		instanceFamilies: neuronInstances,
+		apply:            applyNeuron,
+	}
+
+	mixins = []instanceTypeMixin{
+		NvidiaInstanceTypeMixin,
+		NeuronInstanceTypeMixin,
+	}
+)
+
+// applyInstanceTypeMixins adds the needed OCI hook options to containerd config.toml
+// based on the instance family
+func applyInstanceTypeMixins(cfg *api.NodeConfig, containerdConfig *[]byte) error {
+	for _, mixin := range mixins {
+		if mixin.matches(cfg) {
+			if err := mixin.apply(containerdConfig); err != nil {
+				return err
+			}
+			return nil
+		}
+	}
+	zap.L().Info("No containerd OCI configuration needed..", zap.String("instanceType", cfg.Status.Instance.Type))
+	return nil
+}
+
+type CommandExecutor interface {
+	CombinedOutput(name string, arg ...string) ([]byte, error)
+}
+
+type RealCommandExecutor struct{}
+
+func (r RealCommandExecutor) CombinedOutput(name string, arg ...string) ([]byte, error) {
+	if isAllowedCommand(name, arg...) {
+		cmd := exec.Command(name, arg...)
+		return cmd.CombinedOutput()
+	}
+	return nil, errors.New("unrecognised command")
+
+}
+
+type FileWriter interface {
+	WriteFileWithDir(filePath string, data []byte, perm fs.FileMode) error
+}
+
+type RealFileWriter struct{}
+
+func (r RealFileWriter) WriteFileWithDir(filePath string, data []byte, perm fs.FileMode) error {
+	return util.WriteFileWithDir(filePath, data, perm)
+}
+
+var execCommand CommandExecutor = RealCommandExecutor{}
+var fileWriter FileWriter = RealFileWriter{}
+var nvidiaCtkCommand = "/usr/bin/nvidia-ctk"
+var nvidiaCtkParams = []string{"--quiet", "runtime", "configure", "--runtime=containerd", "--set-as-default", "--dry-run"}
+
+func isAllowedCommand(command string, args ...string) bool {
+	return command == "/usr/bin/nvidia-ctk" && reflect.DeepEqual(args, nvidiaCtkParams)
+}
```

Comment:
This is a bit odd. We don't need a generic `CommandExecutor` interface if it can only do one thing 😉

That interface (and it's impls) shouldn't be public either, AFAICT.

Reply:
I added the `CommandExecutor` so that I can unit test the code more easily(just passing mock command in the test).